From e2cd5c7efa1373faf63eeff454d4e55d9b8bb6ff Mon Sep 17 00:00:00 2001
From: Larisa Grigore <larisa.grigore@nxp.com>
Date: Wed, 20 Jan 2021 13:17:07 +0200
Subject: [PATCH 57/76] s32g: mmu setup: Correct TLB size

Taking into account that now the DTB is placed
before U-Boot text area, the TLB area upper limit
should be CONFIG_DTB_SRAM_ADDR and not CONFIG_SYS_TEXT_BASE.

Issue: ALB-6137
Signed-off-by: Larisa Grigore <larisa.grigore@nxp.com>
---
 arch/arm/cpu/armv8/s32/cpu.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/arch/arm/cpu/armv8/s32/cpu.c b/arch/arm/cpu/armv8/s32/cpu.c
index 0b7acc9b39..c589f34724 100644
--- a/arch/arm/cpu/armv8/s32/cpu.c
+++ b/arch/arm/cpu/armv8/s32/cpu.c
@@ -193,11 +193,20 @@ static void enable_snooping(void)
 }
 #endif
 
+static unsigned long get_tlb_size(void)
+{
+#ifdef CONFIG_S32V234
+	return CONFIG_SYS_TEXT_BASE - S32_IRAM_MMU_TABLES_BASE;
+#else
+	return CONFIG_DTB_SRAM_ADDR - S32_IRAM_MMU_TABLES_BASE;
+#endif
+}
+
 static inline void early_mmu_setup(void)
 {
 	/* global data is already setup, no allocation yet */
 	gd->arch.tlb_addr = S32_IRAM_MMU_TABLES_BASE;
-	gd->arch.tlb_size = CONFIG_SYS_TEXT_BASE - S32_IRAM_MMU_TABLES_BASE;
+	gd->arch.tlb_size = get_tlb_size();
 
 #if defined(CONFIG_S32_SKIP_RELOC) && !defined(CONFIG_S32_ATF_BOOT_FLOW)
 	sram_clr(gd->arch.tlb_addr, gd->arch.tlb_size);
@@ -235,7 +244,7 @@ static inline void final_mmu_setup(void)
 	/* global data is already setup, no allocation yet */
 	gd->arch.tlb_addr = S32_SDRAM_MMU_TABLES_BASE;
 	gd->arch.tlb_fillptr = gd->arch.tlb_addr;
-	gd->arch.tlb_size = CONFIG_SYS_TEXT_BASE - S32_IRAM_MMU_TABLES_BASE;
+	gd->arch.tlb_size = get_tlb_size();
 
 #if defined(CONFIG_S32_SKIP_RELOC) && !defined(CONFIG_S32_ATF_BOOT_FLOW)
 	sram_clr(gd->arch.tlb_addr, gd->arch.tlb_size);
-- 
2.17.1

